pyroute2/netlink/diag/__init__.py

```diff
@@ -256,11 +281,12 @@ def get_sock_stats(self,
             req = unix_diag_req()
             req['udiag_states'] = states
             req['udiag_show'] = show
-        elif family == AF_INET:
+        elif family == AF_INET or family == AF_INET6:
             req = inet_diag_req()
             req['idiag_states'] = states
             req['sdiag_protocol'] = protocol
             req['idiag_ext'] = extensions
+            self.marshal.adapt(protocol)
```

Comment:
This approach is not thread-safe — we may end up parsing messages with a wrong decoding routine, when there will be more than one marshal.adapt.

Personally I prefer a message based choice, look at the code here: https://github.com/svinota/pyroute2/blob/master/pyroute2/netlink/rtnl/ifinfmsg/__init__.py#L615

But it works only on the NLA level, I didn't adapt it to the message level yet.

What may be here instead of `tcp_inet_diag_msg`? Is there any UDP-specific structure, maybe? Could you please point me into the relevant kernel code?

Reply:
I see. Ah, yes, this callback way would be it, I was wondering already. First thought about inheritance, but there's no way without dynamics.

No, except for SCTP in addition to TCP, I didn't find anything yesterday that's filling INET_DIAG_INFO. (ref: https://elixir.bootlin.com/linux/latest/source/net/sctp/sctp_diag.c#L117) E.g. UDP, RAW are only updating sk_mem_info in their specific way. Though, was only checking superficially, might be more ...